### PR TITLE
refactor(obligations): Make separate database tables for classification and type

### PR DESCRIPTION
**Pull request by deo002**
## Changes

- Make separate database tables for classification and type. They are not a fixed set of values. We can add/remove the possible classification and type for obligation. Support for this functionality will be added in future PRs.
- Use GORM notation for many-to-many relationship of obligation maps.
- While converting from the new Obligation struct to json, the classification, type and shortnames were not being marshalled in the desired format. So wrote custom marshaller and unmarshaller for it. We could have used Converter functions too, same effort would be required.
- Added BeforeCreate and BeforeUpdate hook for data validation(non empty strings for topic, text, unique topic and text, valid classification and type values etc).
- Deleted optional data type wrappers as they are no longer needed.

## Submitter Checklist

- [ ] Includes tests (if there is a feature changed/added)
- [X] Includes docs ( if changes are user facing)
- [X] I have tested my changes locally.

### CODE REVIEW

**Review by GMishx — APPROVED** [review]
Changes looks good. Needs test

File: pkg/api/licenses.go [hunk 1]
```diff
@@ -913,10 +913,14 @@ func ImportLicenses(c *gin.Context) {
 			errMessage, importStatus, oldLicense, newLicense := utils.InsertOrUpdateLicenseOnImport(tx, &licenses[i], &externalRefs[i])
 
 			if importStatus == utils.IMPORT_FAILED {
+				erroredLicense := ""
+				if licenses[i].Shortname != nil {
+					erroredLicense = *licenses[i].Shortname
+				}
```
Comment (avinal) [inline pkg/api/licenses.go:919]: Good one

File: pkg/models/types.go [hunk 2]
```diff
@@ -271,18 +279,260 @@ type AuditResponse struct {
...
+	TextUpdatable              *bool                     `gorm:"not null;default:false"`
+	Md5                        string                    `gorm:"unique;not null"`
+	ObligationClassificationId int64                     `gorm:"not null"`
+	ObligationTypeId           int64                     `gorm:"not null"`
+	Licenses                   []*LicenseDB              `gorm:"many2many:obligation_licenses;"`
+	Type                       *ObligationType           `gorm:"foreignKey:ObligationTypeId"`
+	Classification             *ObligationClassification `gorm:"foreignKey:ObligationClassificationId"`
+}
+
+func (o *Obligation) BeforeCreate(tx *gorm.DB) (err error) {
```
Comment (avinal) [inline pkg/models/types.go:310]: I would highly recommend moving all functions out of this file. Probably in the same package but a different file. It somewhat undermines the structure of the project.

Not a blocker.

**Review by avinal — APPROVED** [review]
Few nitpicks, looks good to me.

**Comment by GMishx** [comment]
For existing database, these new columns should hold some default values:

```
LicenseDb/laas -dbname laas -datafile fossology/install/db/licenseRef.json -populatedb true

2024/10/18 10:33:40 LicenseDb/cmd/laas/main.go:53 ERROR: column "obligation_classification_id" of relation "obligations" contains null values (SQLSTATE 23502)
[0.252ms] [rows:0] ALTER TABLE "obligations" ADD "obligation_classification_id" bigint NOT NULL
2024/10/18 10:33:40 Failed to automigrate database: ERROR: column "obligation_classification_id" of relation "obligations" contains null values (SQLSTATE 23502)

Process finished with the exit code 1
```

File: pkg/models/types.go (hunk 2 above)
Comment (deo002) [reply pkg/models/types.go:310]: Yes. What do you think about moving a GORM struct and all it's methods into separate files? Like one for licenses, one for obligations etc.
I'll do it in another PR, this one's already big.

Comment (avinal) [reply pkg/models/types.go:310]: This is a good idea, under `models` package, you can separate them into differnt files.

**Comment by GMishx** [comment]
While trying to pass wrong data, the unmarshling failed which printed some text making the overall response invalid json.

```json
can't parse JSON.  Raw result:

{"status":500,"message":"Failed to update license","error":"obligation classification must be one of the following values: GREEN WHITE YELLOW RED","path":"/api/v1/obligations/provide license upstream","timestamp":"2024-10-22T10:56:59+05:30"}{"status":200,"data":[{"topic":null,"type":"RISK","text":null,"classification":"WHITTTE","modifications":null,"comment":null,"active":null,"text_updatable":null,"shortnames":[]}],"paginationmeta":{"resource_count":1}}
```

**Review by GMishx — APPROVED** [review]
Tested, working as expected.

**Pull request merged**